File: pjecalc-fonte/negocio/br/jus/trt8/pjecalc/negocio/dominio/calculo/multa/TotalizadorDeMulta.java
/*
 * Decompiled with CFR 0.152.
 */
package br.jus.trt8.pjecalc.negocio.dominio.calculo.multa;

import br.jus.trt8.pjecalc.base.comum.Utils;
import br.jus.trt8.pjecalc.negocio.comum.Total;
import br.jus.trt8.pjecalc.negocio.dominio.calculo.Calculo;
import br.jus.trt8.pjecalc.negocio.dominio.calculo.multa.Multa;
import java.io.Serializable;
import java.math.BigDecimal;

public class TotalizadorDeMulta
implements Serializable {
    private static final long serialVersionUID = 1L;
    private Calculo calculo;
    private boolean isCalculado;
    private Total reclamanteReclamado;
    private Total reclamadoReclamante;
    private Total terceiroReclamado;

    public TotalizadorDeMulta(Calculo calculo) {
        this.calculo = calculo;
        this.reset();
    }

    public void reset() {
        this.isCalculado = false;
    }

    private TotalizadorDeMulta calcular() {
        if (!this.isCalculado) {
            this.reclamanteReclamado = Total.newInstance(true);
            this.reclamadoReclamante = Total.newInstance(true);
            this.terceiroReclamado = Total.newInstance(true);
            for (Multa multa : this.calculo.getMultasDoCalculo()) {
                if (!Utils.naoNulo(multa.getValorTotal())) continue;
                switch (multa.getTipoCredorDevedor()) {
                    case RECLAMANTE_RECLAMADO: {
                        this.reclamanteReclamado.acumular(multa.getValorTotal());
                        break;
                    }
                    case RECLAMADO_RECLAMANTE: {
                        this.reclamadoReclamante.acumular(multa.getValorTotal());
                        break;
                    }
                    case TERCEIRO_RECLAMADO: {
                        this.terceiroReclamado.acumular(multa.getValorTotal());
                        break;
                    }
                }
            }
            this.isCalculado = true;
        }
        return this;
    }

    public BigDecimal getTotalTipoReclamanteReclamado() {
        return this.calcular().reclamanteReclamado.getValor();
    }

    public BigDecimal getTotalTipoReclamadoReclamante() {
        return this.calcular().reclamadoReclamante.getValor();
    }

    public BigDecimal getTotalTipoTerceiroReclamado() {
        return this.calcular().terceiroReclamado.getValor();
    }
}

File: src/lib/pjecalc/core/dominio/calculo/multa/totalizador-de-multa.ts
/**
 * PJe-Calc v2.15.1 — TotalizadorDeMulta
 * Porte 1:1 de: br.jus.trt8.pjecalc.negocio.dominio.calculo.multa.TotalizadorDeMulta
 *
 * Ref Java: pjecalc-fonte/.../calculo/multa/TotalizadorDeMulta.java
 *
 * Soma multas por categoria CredorDevedor (reclamanteReclamado,
 * reclamadoReclamante, terceiroReclamado). `reset()` invalida cache.
 */
import Decimal from 'decimal.js';
import type { Calculo } from '../calculo';
import { CredorDevedorMultaEnum } from '../../../constantes/enums';
import type { Multa } from './multa';

const ZERO = new Decimal(0);

class Total {
  private valor: Decimal = ZERO;
  acumular(v: Decimal | null): void { if (v !== null) this.valor = this.valor.plus(v); }
  getValor(): Decimal { return this.valor; }
}

export class TotalizadorDeMulta {
  private calculo: Calculo;
  private isCalculado = false;
  private reclamanteReclamado = new Total();
  private reclamadoReclamante = new Total();
  private terceiroReclamado = new Total();

  constructor(calculo: Calculo) {
    this.calculo = calculo;
  }

  reset(): void { this.isCalculado = false; }

  private calcular(): this {
    if (this.isCalculado) return this;
    this.reclamanteReclamado = new Total();
    this.reclamadoReclamante = new Total();
    this.terceiroReclamado = new Total();
    const calcExt = this.calculo as unknown as { getMultasDoCalculo?(): Iterable<Multa> };
    for (const m of calcExt.getMultasDoCalculo?.() ?? []) {
      const total = m.getValorTotal();
      if (total === null) continue;
      switch (m.getTipoCredorDevedor()) {
        case CredorDevedorMultaEnum.RECLAMANTE_RECLAMADO:
          this.reclamanteReclamado.acumular(total);
          break;
        case CredorDevedorMultaEnum.RECLAMADO_RECLAMANTE:
          this.reclamadoReclamante.acumular(total);
          break;
        case CredorDevedorMultaEnum.TERCEIRO_RECLAMADO:
          this.terceiroReclamado.acumular(total);
          break;
      }
    }
    this.isCalculado = true;
    return this;
  }

  getTotalTipoReclamanteReclamado(): Decimal { return this.calcular().reclamanteReclamado.getValor(); }
  getTotalTipoReclamadoReclamante(): Decimal { return this.calcular().reclamadoReclamante.getValor(); }
  getTotalTipoTerceiroReclamado(): Decimal { return this.calcular().terceiroReclamado.getValor(); }
}
